Support Azure storage container operations

yaxia opened the pull request:
Add support for API:
``` ruby
  create_container
  delete_container 
  list_containers 
  get_container_properties 
  get_container_acl 
```

shaffan-chaudhry-confiz commented on lib/fog/azurerm/storage.rb:
```diff
@@ -32,20 +46,36 @@ def initialize(_options = {})
+            @storage_mgmt_client.subscription_id = options[:subscription_id]
+          end
+
+          if Fog::Credentials::AzureRM.new_account_credential? options
```
You should not be writing this code here. This functionality should be moved to the individual requests as we did here: https://github.com/fog/fog-azure-rm/blob/master/lib/fog/azurerm/requests/storage/delete_disk.rb

yaxia replied:
Thanks for reviewing.

Let me explain more about the reason I am adding code here. In the Azure Storage REST API, there are two categories of APIs:

- **Resource management APIs**: Targeting on creating/deleting/updating/fetching storage accounts
- **Storage data APIs**: Targeting on the storage blob/table/queue/file operations

With this understanding, I assume there are two scenarios when users create a `Fog::Storage::AzureRM` instance. One is for managing the storage account and the other is for managing the blob/table/queue/file resources. 

For the Azure storage data operations, there could be always be intensive calls. So I am considering optimizing the calls:
- Reducing REST calls. I may not get the blob client as the code in `delete_disk` does, which makes an additional call to list the accounts and searches for the account key.
- Reuse the client instance. As the intensive calls against the storage account, I'd like to reuse the client instance. That's why I create a instance variable `@blob_client` in the `Fog::Storage::AzureRM`. I will also add `@table_client`, `@queue_client` and `@file_client` in the similar way, if you agree with this approach.

On the other hand, it is a common practice that the user is not a subscription admin and he/she doesn't have a client ID and subscription ID. What the user has are the storage account name and key or a SAS token. When the user creates a `Fog::Storage::AzureRM` instance, getting the data client from the object he/she gets looks an intuitive way.

Does it make sense to you?

samawia-moin-confiz replied:
Yeah. It make sense. Its fine to make the blob client object in the storage service.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/container.rb:
```diff
@@ -0,0 +1,35 @@
+      # This class is giving implementation of create and delete a container.
+      class Container < Fog::Model
+        identity  :name
+        attribute :properties
```
You should replace 'properties' with following attributes i.e
*last_modified*
*etag*
*lease_status*
*lease_state*

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/container.rb:
```diff
@@ -0,0 +1,35 @@
+
+        def create(options = {})
+          requires :name
+          merge_attributes(service.create_container name, options)
```
After replacing properties with those 4 attributes you will parse the response coming from *service.create_container* and then call *merge_attributes* passing that parsed response as a parameter.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/requests/storage/delete_container.rb:
```diff
@@ -0,0 +1,27 @@
+      end
+    end
+  end
+end
```
Please fix all hound violations.

yaxia replied:
This is actually fixed. Hound-CI somehow doesn't close the item.
The final

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/container.rb:
```diff
@@ -0,0 +1,35 @@
+          merge_attributes(service.get_container_properties name, options)
+        end
+
+        def get_acl(options = {})
```
Please replace *acl* with some meaning full name.

yaxia replied:
I've renamed the acl to access_control_list.

shaffan-chaudhry-confiz commented:
Please resolve all the issues mentioned in inline comments.

Requests unit tests are missing. Please write unit tests of all the requests which you have created:
*create_container*
*delete_container*
*list_containers*
*get_container_properties*
*get_container_acl*

yaxia commented on test/models/storage/test_containers.rb:
```diff
@@ -0,0 +1,41 @@
+      end
+    end
+  end
+end
```
This is actually fixed. Hound-CI somehow doesn't close the item.

It is 42 lines in fact. 
https://github.com/yaxia/fog-azure-rm/blob/master/test/models/storage/test_containers.rb

yaxia commented on test/models/storage/test_container.rb:
```diff
@@ -0,0 +1,63 @@
+      assert @container.delete
+    end
+  end
+end
```
This is actually fixed. Hound-CI somehow doesn't close the item.
It is 68 lines in fact. 
https://github.com/yaxia/fog-azure-rm/blob/master/test/models/storage/test_container.rb

yaxia commented on lib/fog/azurerm/requests/storage/list_containers.rb:
```diff
@@ -0,0 +1,47 @@
+      end
+    end
+  end
+end
```
This is actually fixed. Hound-CI somehow doesn't close the item.
It is 50 lines in fact. 
https://github.com/yaxia/fog-azure-rm/blob/master/lib/fog/azurerm/requests/storage/list_containers.rb

yaxia commented on lib/fog/azurerm/requests/storage/get_container_properties.rb:
```diff
@@ -0,0 +1,36 @@
+      end
+    end
+  end
+end
```
This is actually fixed. Hound-CI somehow doesn't close the item.
It is 37 lines in fact. 
https://github.com/yaxia/fog-azure-rm/blob/master/lib/fog/azurerm/requests/storage/get_container_properties.rb

yaxia commented:
@shaffan-chaudhry-confiz Thanks for reviewing the long list of the files.

I've addressed all the comments. The unclosed hound comments seems not valid because it doesn't show the latest change and the missing lines are added.

I also added the unit test cases for the requests. Please be noted that I didn't directly call any REST API for storage data plane. They are wrapped in the azure-storage gem and I use the gem instead. So I didn't make any request in the pattern of promise and check the response.

samawia-moin-confiz commented on lib/fog/azurerm/credentials.rb:
```diff
@@ -19,19 +19,26 @@ def self.get_new_credentials(tenant_id, client_id, client_secret)
+      def self.new_account_credential?(options = {})
+        @account_name != options[:azure_storage_account_name] ||
+          @account_key != options[:azure_storage_access_key] ||
+          @connection_string != options[:azure_storage_connection_string]
```
There can be azure_storage_account_name and azure_storage_access_key only. azure_storage_connection_string is not necessary.

yaxia replied:
azure_storage_connection_string can useful for specifying the default protocol or user defined endpoint.
https://azure.microsoft.com/en-us/documentation/articles/storage-configure-connection-string/

samawia-moin-confiz commented on lib/fog/azurerm/storage.rb:
```diff
@@ -3,10 +3,17 @@ module Fog
+      recognizes :azure_storage_account_name
+      recognizes :azure_storage_access_key
+      recognizes :azure_storage_connection_string
+      recognizes :debug
```
Attributes required for blob service, should not be class attributes.In this case, user will have to provide extra attributes every time when these attributes are not required and just want to create a storage account.

yaxia replied:
Agree, so I use "recognizes" instead of "requires" to register the class attributes. User can choose use the parameter set for storage accounts or storage data. They are not required in creating the Fog::Storage::AzureRM instance.
Both the following calls should work.
```ruby
# Creating an instance to operate on storage accounts
azure_storage_service = Fog::Storage.new(
  :provider => 'AzureRM',
  :tenant_id => '<Tenantid>',     
  :client_id =>    '<Clientid>', 
  :client_secret => '<ClientSecret>',
  :subscription_id => '<Subscriptionid>'
)
```
or
```ruby
# Creating an instance to operate on storage data
azure_storage_service = Fog::Storage.new(
  :provider => 'AzureRM',
  :azure_storage_account_name => '<account_name>',
  :azure_storage_access_key =>    '<account_key>'
)
```

shaffan-chaudhry-confiz commented:
Hound is using Rubocop for static code analysis on the code. You can use rubocop gem to resolve all violations on your local machine.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/container.rb:
```diff
@@ -0,0 +1,50 @@
+          service.delete_container name, options
+        end
+
+        def self.flatten_properties(container)
```
Please change this function name to 'parse' instead of 'flatten_properties'. You may have seen that we have followed that naming convention in all the models.

yaxia replied:
Changed. Thanks.

shaffan-chaudhry-confiz commented:
Please add documentation of the features you added. I will appreciate if you follow documenting convention.
Here is the file where you will be adding docs: https://github.com/fog/fog-azure-rm/blob/master/lib/fog/azurerm/docs/storage.md

yaxia commented:
Revised. Please review the changes in READMD.md. I added the docs for container operation and made some changes for the style convention on the existing doc:
- Indent 2 spaces instead of 4
- Align the code lines

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/container.rb:
```diff
@@ -0,0 +1,50 @@
+        end
+
+        def self.parse(container)
+          return container unless container.key?('properties')
```
Your code is breaking at this line #38: undefined method 'key?' 
Because you are getting Azure::Storage::Blob::Container::Container class instance in response, not a hash.

yaxia replied:
Good catch! Thanks.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/containers.rb:
```diff
@@ -0,0 +1,53 @@
+          service.list_containers(options).each do |container|
+            hash = {}
+            if container.is_a? Hash
+              hash['name'] = container['name']
```
You should reuse 'parse' method of container class to remove code repetition.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/models/storage/container.rb:
```diff
@@ -0,0 +1,50 @@
+          merge_attributes(service.get_container_access_control_list(name, options)[0])
+        end
+
+        def delete(options = {})
```
Change this function name to 'destroy' instead of 'delete'. Its our convention.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/docs/storage.md:
````diff
@@ -36,57 +63,104 @@ Check Storage Account Name Availability.This operation checks that account name
+Create a storage container in the current storage account.
+
+```ruby
+container = azure_storage_service.create_container('<container name>')
````
You should be following this convention here:

```
storage.containers.create(
  name: '<container name>'
)
```

shaffan-chaudhry-confiz commented on lib/fog/azurerm/docs/storage.md:
````diff
@@ -36,57 +63,104 @@ Check Storage Account Name Availability.This operation checks that account name
+List all the storage containers in the current storage accounts.
+
+```ruby
+containers = azure_storage_service.create_container('<container name>')
````
You should be following this convention here:

`containers = azure_storage_service.containers`

shaffan-chaudhry-confiz commented on lib/fog/azurerm/docs/storage.md:
````diff
@@ -36,57 +63,104 @@ Check Storage Account Name Availability.This operation checks that account name
+Get the storage container properties. The properties will not fetch the access control list. Call `get_container_access_control_list` to fetch it.
+
+```ruby
+properties = azure_storage_service.get_container_properties('<container name>')
````
You should be following this convention here:

```
container = azure_storage_service.containers.get('<container name>')
properties = container.get_properties
```

yaxia replied:
It is workable. But my question here is: 
```ruby
container = azure_storage_service.containers.get('<container name>')
properties = container.get_properties 
```
perform two requests to the server (the first one to list the containers and the second one to get the container properties) where only one `get_container_properties` API can make it.

I am OK to follow the convention, just have performance concern.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/docs/storage.md:
````diff
@@ -36,57 +63,104 @@ Check Storage Account Name Availability.This operation checks that account name
+Get the permissions for the specified container. The permissions indicate whether container data may be accessed publicly.
+
+```ruby
+access_control_list = azure_storage_service.get_container_access_control_list('<container name>')
````
You should be following this convention here:

```
container = azure_storage_service.containers.get('<container name>')
access_control_list = container.get_access_control_list
```

yaxia replied:
The same question as the above one.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/docs/storage.md:
````diff
@@ -36,57 +63,104 @@ Check Storage Account Name Availability.This operation checks that account name
+Mark the specified container for deletion. The container and any blobs contained within it are later deleted during garbage collection.
+
+```ruby
+result = azure_storage_service.delete_container('<container name>')
````
You should be following this convention here:

```
container = azure_storage_service.containers.get('<container name>')
result = container.destroy
```

yaxia replied:
The same question as the above one.

shaffan-chaudhry-confiz commented on lib/fog/azurerm/requests/storage/delete_container.rb:
```diff
@@ -0,0 +1,26 @@
+    class AzureRM
+      # This class provides the actual implemention for service calls.
+      class Real
+        def delete_container(name, options = {})
```
This function should return true if container is deleted successfully.

yaxia replied:
Good catch!

abbas-sheikh-confiz commented on lib/fog/azurerm/requests/storage/get_container_access_control_list.rb:
```diff
@@ -0,0 +1,29 @@
+        def get_container_access_control_list(name, options = {})
+          Fog::Logger.debug "Get container ACL: #{name}."
+          begin
+            container = @blob_client.get_container_acl(name, options)
```
Variable should be renamed to container_acl

abbas-sheikh-confiz commented on lib/fog/azurerm/requests/storage/get_container_properties.rb:
```diff
@@ -0,0 +1,36 @@
+        def get_container_properties(name, options = {})
+          Fog::Logger.debug "Get container properties: #{name}."
+          begin
+            container = @blob_client.get_container_properties(name, options)
```
variable should be renamed to container_properties
